google/cloud/internal/oauth2_external_account_credentials_test.cc

```diff
@@ -0,0 +1,492 @@
+    ]
+  }
+})""";
+
+std::unique_ptr<RestResponse> MakeMockResponseError() {
+  auto response = absl::make_unique<MockRestResponse>();
+  EXPECT_CALL(*response, StatusCode)
+      .WillRepeatedly(Return(HttpStatusCode::kNotFound));
+  EXPECT_CALL(std::move(*response), ExtractPayload)
+      .Times(AtMost(1))
+      .WillRepeatedly([] {
+        auto payload = absl::make_unique<MockHttpPayload>();
+        // This is shared by the next two mocking functions.
+        auto c = std::make_shared<std::string>(kErrorPayload);
+        EXPECT_CALL(*payload, HasUnreadData).WillRepeatedly([c] {
+          return !c->empty();
+        });
+        EXPECT_CALL(*payload, Read)
+            .WillRepeatedly([c](absl::Span<char> buffer) {
+              auto const n = (std::min)(buffer.size(), c->size());
+              std::copy(c->begin(), std::next(c->begin(), n), buffer.begin());
+              c->assign(c->substr(n));
+              return n;
+            });
+        return std::unique_ptr<rest_internal::HttpPayload>(std::move(payload));
+      });
+  return response;
+}
+
+using FormDataType = std::vector<std::pair<std::string, std::string>>;
+
+TEST(ExternalAccount, Working) {
+  auto const test_url = std::string{"https://sts.example.com/"};
+  auto const expected_access_token = std::string{"test-access-token"};
+  auto const expected_expires_in = std::chrono::seconds(3456);
+  auto const json_response = nlohmann::json{
+      {"access_token", expected_access_token},
+      {"expires_in", expected_expires_in.count()},
+      {"issued_token_type", "urn:ietf:params:oauth:token-type:access_token"},
+      {"token_type", "Bearer"},
+  };
+  auto mock_source = [](Options const&) {
+    return make_status_or(internal::SubjectToken{"test-subject-token"});
+  };
+  auto const info = ExternalAccountInfo{
+      "test-audience",
+      "test-subject-token-type",
+      test_url,
+      mock_source,
+  };
+  MockClientFactory client_factory;
+  EXPECT_CALL(client_factory, Call).WillOnce([&]() {
+    auto mock = absl::make_unique<MockRestClient>();
+    EXPECT_CALL(*mock, Post(_, An<FormDataType const&>()))
+        .WillOnce([test_url, info, response = json_response.dump()](
+                      RestRequest const& request, auto const& form_data) {
+          EXPECT_THAT(
+              form_data,
+              UnorderedElementsAre(
+                  Pair("grant_type",
+                       "urn:ietf:params:oauth:grant-type:token-exchange"),
+                  Pair("requested_token_type",
+                       "urn:ietf:params:oauth:token-type:access_token"),
+                  Pair("scope",
+                       "https://www.googleapis.com/auth/cloud-platform"),
+                  Pair("audience", "test-audience"),
+                  Pair("subject_token_type", "test-subject-token-type"),
+                  Pair("subject_token", "test-subject-token")));
+          EXPECT_EQ(request.path(), test_url);
+          return MakeMockResponseSuccess(response);
```

Comment:
nit: std::move(response) ?

Reply:
Hmm... that requires making the lambda `mutable`, but since it is only called once, I will allow it.